editor/store/selectors.js

```diff
@@ -1204,141 +1229,170 @@ export function getNotices( state ) {
- * @property {boolean}  isDisabled        Whether or not the user should be prevented from inserting this item.
- */
-
-/**
- * Given a regular block type, constructs an item that appears in the inserter.
- *
- * @param {Object}           state             Global application state.
- * @param {string[]|boolean} allowedBlockTypes Allowed block types, or true/false to enable/disable all types.
- * @param {Object}           blockType         Block type, likely from getBlockType().
- *
- * @return {Editor.InserterItem} Item that appears in inserter.
- */
-function buildInserterItemFromBlockType( state, allowedBlockTypes, blockType ) {
-	if ( ! allowedBlockTypes || ! blockType ) {
-		return null;
-	}
-
-	const blockTypeIsDisabled = Array.isArray( allowedBlockTypes ) && ! includes( allowedBlockTypes, blockType.name );
-	if ( blockTypeIsDisabled ) {
-		return null;
-	}
-
-	if ( blockType.isPrivate ) {
-		return null;
-	}
-
-	return {
-		id: blockType.name,
-		name: blockType.name,
-		initialAttributes: {},
-		title: blockType.title,
-		icon: blockType.icon,
-		category: blockType.category,
-		keywords: blockType.keywords,
-		isDisabled: !! blockType.useOnce && getBlocks( state ).some( ( block ) => block.name === blockType.name ),
-	};
-}
+ * @property {boolean}  isDisabled        Whether or not the user should be prevented from inserting
+ *                                        this item.
+ * @property {number}   utility           How useful we think this item is, between 0 and 3.
+ * @property {number}   frecency          Hueristic that combines recency and frecency. Useful for
+ *                                        ordering items by relevancy.
+ */
+export const getInserterItems = createSelector(
+	( state, editorAllowedBlockTypes, parentUID = null ) => {
+		const editorAllowList = parseAllowList( editorAllowedBlockTypes );
+
+		let parentAllowList, parentName;
+		if ( parentUID ) {
+			const parentBlockListSettings = getBlockListSettings( state, parentUID );
+			const parentAllowedBlockTypes = get( parentBlockListSettings, [ 'supportedBlocks' ] );
+			parentAllowList = parseAllowList( parentAllowedBlockTypes );
+			parentName = getBlockName( state, parentUID );
+		} else {
+			parentAllowList = parseAllowList( true );
+			parentName = getCurrentPostType( state );
+		}
 
-/**
- * Given a shared block, constructs an item that appears in the inserter.
- *
- * @param {Object}           state             Global application state.
- * @param {string[]|boolean} allowedBlockTypes Allowed block types, or true/false to enable/disable all types.
- * @param {Object}           sharedBlock       Shared block, likely from getSharedBlock().
- *
- * @return {Editor.InserterItem} Item that appears in inserter.
- */
-function buildInserterItemFromSharedBlock( state, allowedBlockTypes, sharedBlock ) {
-	if ( ! allowedBlockTypes || ! sharedBlock ) {
-		return null;
-	}
+		const calculateUtility = ( category, count, isContextual ) => {
```

Comment:
If I understand properly the utility is essentially usefull when displaying the "suggested" category right?

Order the items in the inserter: Suggested > frecent > common

And it's also used to exclude items from the Suggestions (utility = 0).

I think it's a bit confusing (maybe just because it's a new API), what if we don't filter at all, if we splice later (seems like we do), we don't need to filter.

Reply:
Yep, the behaviour of the _Suggested_ tab is now to show items in the following order:

1. Blocks that are contextually useful. For example, if you are inserting into a _Product_ block, _Add to Cart_ is contextually useful.
2. Blocks that have been inserted by the current user in the past.
3. Blocks that are in the common category.
4. All other blocks.

I introduced `utility` as a way to achieve this, since the inserter logic was growing quite complex, e.g. `fillWithCommonBlocks` was quite difficult to reason about.